core/src/main/java/com/graphhopper/coll/GHTreeMapComposed.java

```diff
@@ -34,39 +35,60 @@
     private final TreeMap<Long, Integer> map;
 
     public GHTreeMapComposed() {
-        map = new TreeMap<>();
+        map = new TreeMap<>(new Comparator<Long>() {
+            //  we cannot just use the long sorting because the values are floats
+            @Override
+            public int compare(Long o1, Long o2) {
+                // for two entries to be equal both value and key must be equal
+                if (o1.equals(o2)) return 0;
+                int value1 = bitUtil.getIntHigh(o1);
+                int value2 = bitUtil.getIntHigh(o2);
+                if (value1 == value2) {
+                    // we enforce a deterministic order by looking at the size of the key (although there is no real
+                    // reason to prefer one entry over the other)
+                    int key1 = bitUtil.getIntLow(o1);
+                    int key2 = bitUtil.getIntLow(o2);
+                    if (key1 == key2) return 0;
+                    return key1 < key2 ? -1 : 1;
+                }
+                float f1 = Float.intBitsToFloat(value1);
+                float f2 = Float.intBitsToFloat(value2);
+                return Float.compare(f1, f2);
+            }
+        });
```

Comment:
Yes, I tried using some fixed precision, but this turned out to be rather instable, because depending on the metric one uses its hard to know the required range beforehand and sometimes the values become very small or very large. I still think it should be possible, but then again I do not think switching to float hurts and its a lot easier to work with.